clippy_lints/src/methods.rs

```diff
@@ -1220,6 +1241,37 @@ fn lint_map_unwrap_or_else<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &'tcx hir
     }
 }
 
+/// lint use of `_.map_or(None, _)` for `Option`s
+fn lint_map_or_none<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &'tcx hir::Expr, map_or_args: &'tcx [hir::Expr]) {
+    // check if the first non-self argument to map_or() is None
+    let map_or_arg_is_none = if let hir::Expr_::ExprPath(ref qpath) = map_or_args[1].node {
```

Comment:
You need to ensure that the type is `Option` before doing this check. Otherwise a random type with a `map_or` function can cause this lint to panic if the `map_or` has zero arguments (so just the self).